# 선형회귀 모델 작성, 예측 ,평가 

In [97]:
import numpy as np
import matplotlib.pyplot as plt 
import pandas as pd 
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
# 데이터를 학습용과 테스트용으로 분리
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
# 평균 제곱 오차 계산
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd 
from sklearn.ensemble import RandomForestRegressor
%matplotlib

# LinearRegressor & RandomForestRegressor()

Using matplotlib backend: module://matplotlib_inline.backend_inline


In [98]:
prdf = pd.read_csv('data/premium.csv')
prdf

age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1338 rows x 7 columns]

In [99]:
from sklearn.preprocessing import LabelEncoder

labelen = LabelEncoder()
prdf['sex'] = labelen.fit_transform(prdf['sex'])
prdf['smoker'] = labelen.fit_transform(prdf['smoker'])
prdf['region'] = labelen.fit_transform(prdf['region'])

In [100]:
prdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   int64  
 2   bmi       1333 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   int64  
 5   region    1338 non-null   int64  
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(5)
memory usage: 73.3 KB


In [101]:
prdf['bmi'] = prdf['bmi'].fillna(prdf['bmi'].mean())
prdf

age  sex     bmi  children  smoker  region      charges
0      19    0  27.900         0       1       3  16884.92400
1      18    1  33.770         1       0       2   1725.55230
2      28    1  33.000         3       0       2   4449.46200
3      33    1  22.705         0       0       1  21984.47061
4      32    1  28.880         0       0       1   3866.85520
...   ...  ...     ...       ...     ...     ...          ...
1333   50    1  30.970         3       0       1  10600.54830
1334   18    0  31.920         0       0       0   2205.98080
1335   18    0  36.850         0       0       2   1629.83350
1336   21    0  25.800         0       0       3   2007.94500
1337   61    0  29.070         0       1       1  29141.36030

[1338 rows x 7 columns]

In [102]:
x= prdf.drop('charges', axis=1).values
y= prdf['charges'].values

In [103]:
prdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   int64  
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   int64  
 5   region    1338 non-null   int64  
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(5)
memory usage: 73.3 KB


In [104]:
xtrain, xtest, ytrain, ytest = train_test_split(x,y,test_size=0.3,random_state=156)
lr = LinearRegression()
lr.fit(xtrain, ytrain)
lr_pred = lr.predict(xtest)
lr_pred[:3]

array([14474.70246359, -1367.94174448, 11182.0795591 ])

In [105]:
# 평가
# 평균 제곱 오차 계산
mse = mean_squared_error(ytest, lr_pred)
# RMSE 계산
rmse = np.sqrt(mse)
R2 = r2_score(ytest, lr_pred)
print(f' mse = {mse}, rmse = {rmse}, R2 = {R2}')

 mse = 34719047.52833967, rmse = 5892.287122021437, R2 = 0.7314050294401666


In [106]:
# 교차 검증
from sklearn.model_selection import cross_val_score
# 평균 제곱 오차 계산
neg_mse_scores = cross_val_score(lr, x, y, scoring='neg_mean_squared_error', cv=5) # neg는 거꾸로
neg_mse_scores

array([-37353966.14780101, -38018280.71475136, -32981193.39000173,
       -39560881.14778336, -37174240.90789752])

In [107]:
# 선형회귀 교차검증 RMSE, R2

# mse, rmse, R2(얘는 따로 계산해야 한다고 함)
RMSE = np.sqrt(neg_mse_scores*-1)
# 결정계수(R2) 계산
lr_score_scores = cross_val_score(lr, x, y, scoring='r2', cv=5)
np.mean(RMSE), np.mean(lr_score_scores) # 선형회귀 RMSE , R2

(np.float64(6081.484710559383), np.float64(0.7461909971637163))

In [108]:
#랜덤포레스트 객체 생성

rf = RandomForestRegressor()
rf.fit(xtrain, ytrain)
rf_pred = rf.predict(xtest)


In [109]:
#랜덤포레스트 교차검증 전

# 평가
# 평균 제곱 오차 계산
mse = mean_squared_error(ytest, rf_pred)
# RMSE 계산
rmse = np.sqrt(mse)
R2 = r2_score(ytest, rf_pred)
print(f' mse = {mse}, rmse = {rmse}, R2 = {R2}')

 mse = 23210767.79759442, rmse = 4817.755472997194, R2 = 0.8204358720331479


In [110]:
from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier

RandomForestRegressor


sklearn.ensemble._forest.RandomForestRegressor

In [111]:
# R2
# 결정계수(R2) 계산
rf_score_scores = cross_val_score(rf, x, y, scoring='r2', cv=5)
neg_mse_scores = cross_val_score(rf, x, y, scoring='neg_mean_squared_error', cv=5) # neg는 거꾸로
rfRMSE = np.sqrt(neg_mse_scores*-1)
# 결정계수(R2) 계산
np.mean(rfRMSE), np.mean(rf_score_scores)

(np.float64(4977.12245719248), np.float64(0.8305709863017057))

In [112]:
neg_mse_scores = cross_val_score(rf, x, y, scoring='neg_mean_squared_error', cv=5) # neg는 거꾸로
rfRMSE = np.sqrt(neg_mse_scores*-1) 

rf_score_scores = cross_val_score(rf, x, y, scoring='r2', cv=5) # R2


np.mean(rfRMSE), np.mean(rf_score_scores)

(np.float64(4940.731795826341), np.float64(0.8299982532489351))

In [113]:
rf_score_scores = cross_val_score(rf, x, y, scoring='r2', cv=5)
# 결정계수(R2) 계산
rf_score_scores, np.mean(rf_score_scores)

(array([0.85214466, 0.76937773, 0.84379162, 0.83464204, 0.85421422]),
 np.float64(0.8308340533660233))

<details>
<summary>결과확인</summary>

### [선형회귀] 
**교차검증 전**   
RMSE = 약 5892   
R2 = 0.73%    

**교차검증 후**   
RMSE 약 6081   
R2 약 0.74%   

### [ 랜덤포레스트 ]   
**교차검증 전**   
RMSE = 약 4767   
R2 = 약 0.82%    

**교차검증 후**   
RMSE = 약 4968   
R2 =  약 0.83%   
</details>

# 다항회귀

y = w0 + w1*x1 + w2*x2 + w3*x3^3

In [117]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression

In [122]:

xtrain, xtest, ytrain, ytest = train_test_split(x,y,test_size=0.3,random_state=156)
model_poly = Pipeline([
    ('poly', PolynomialFeatures(degree=2, include_bias=False)),
    ('Linear',LinearRegression())]
)
model_poly.fit(xtrain, ytrain)

Pipeline(steps=[('poly', PolynomialFeatures(include_bias=False)),
                ('Linear', LinearRegression())])

In [123]:
pred_poly = model_poly.predict(xtest)
mean_squared_error(ytest,pred_poly)

np.float64(21741384.61782171)

In [124]:
r2_score(ytest, pred_poly)

np.float64(0.8318033766166195)

2차~ 5차

In [133]:
results = []

for degrees in range(1,5):
    model_poly = Pipeline([
    ('poly', PolynomialFeatures(degree=degrees, include_bias=False)),
    ('Linear',LinearRegression())]
)
    model_poly.fit(xtrain, ytrain)
    pred_poly = model_poly.predict(xtest)
    mse = mean_squared_error(ytest,pred_poly)
    rmse = np.sqrt(mse)
    r2 = r2_score(ytest, pred_poly)

    results.append({ 'degree': degrees,
                'mse' : mse, 
                'RMSE' : rmse,
                'R2' : r2})

In [134]:
pd.DataFrame(results)

degree           mse         RMSE        R2
0       1  3.471905e+07  5892.287122  0.731405
1       2  2.174138e+07  4662.765769  0.831803
2       3  2.276061e+07  4770.808396  0.823918
3       4  2.992843e+07  5470.687872  0.768466